docs/design.md

```diff
@@ -50,6 +63,328 @@ Default Configuration for Zenoh Sessions:
+* rmw_get_service_names_and_types_by_node
+* rmw_get_client_names_and_types_by_node
+* rmw_get_topic_names_and_types
+
+### Related Zenoh-c APIs
+
+* zc_liveliness_declare_token
+* zc_liveliness_declare_subscriber
+* zc_liveliness_get
+
+## Contexts
+
+A ROS 2 context describes a certain middleware configuration, which can contain 0 or more ROS 2 nodes.
+In `rmw_zenoh_cpp`, a context maps to a Zenoh session, along with a liveliness token for the graph cache and some additional metadata.
+
+### Related RMW APIs
+
+* rmw_get_zero_initialized_init_options
+* rmw_init_options_copy
+* rmw_init_options_fini
+* rmw_get_zero_initialized_context
+* rmw_init
+* rmw_shutdown
+* rmw_context_init
+* rmw_create_guard_condition
+* rmw_destroy_guard_condition
+* rmw_trigger_guard_condition
+* rmw_create_wait_set
+* rmw_destroy_wait_set
+* rmw_wait
+
+### Related Zenoh-c APIs
+
+* zc_liveliness_declare_subscriber
+* zc_liveliness_get
+* z_open
+* z_close
+* z_undeclare_subscriber
+* z_call
+* z_session_check
+
+## Nodes
+
+A ROS 2 node can be though of as the "unit of computation" in a ROS 2 graph; usually one node performs one particular task.
+Nodes may contain publishers, subscriptions, service servers, service clients, action servers, action clients, parameters, and anything else needed to do some computation.
+Zenoh has no conceptual equivalent to the ROS 2 node, so `rmw_zenoh_cpp` creates no Zenoh entities when nodes are created.
+When a new node is created through the RMW API, a liveliness token of type `NN` is sent out.
+
+### Related RMW APIs
+
+* rmw_create_node
+* rmw_destroy_node
+* rmw_node_get_graph_guard_condition
+
+### Related Zenoh-c APIs
+
+* zc_liveliness_declare_token
+
+## Publishers
+
+A ROS 2 publisher sends data to 0 or more connected subscriptions.
+A Zenoh publisher does exactly the same thing, so ROS 2 publishers are mapped onto Zenoh publishers in `rmw_zenoh_cpp`.
+When a new publisher is created, a liveliness token of type `MP` is sent out.
+
+### Related RMW APIs
+
+* rmw_create_publisher
+* rmw_destroy_publisher
+* rmw_publish
+* rmw_publish_serialized_message
+* rmw_borrow_loaned_message
+* rmw_return_loaned_message
+* rmw_publisher_wait_for_all_acked
+* rmw_publisher_get_network_flow_endpoints
+* rmw_publisher_event_init
+
+### Related Zenoh-c APIs
+
+* zc_liveliness_declare_token
+* zc_publish_put_owned
+* z_declare_publisher
+* z_undeclare_publisher
+* z_publisher_put
+
+## Subscriptions
+
+A ROS 2 subscription receives data from 1 or more connected publishers.
+A Zenoh subscriber does exactly the same thing, so ROS 2 subscriptions are mapped onto Zenoh subscribers in `rmw_zenoh_cpp`.
+When new data arrives, a callback within `rmw_zenoh_cpp` is executed, which takes ownership of the data and signals that there is data available.
+Then rmw_wait can find out that there is data available, and the data can be delivered via rmw_take.
+When a new subscription is created, a liveliness token of type `MS` is sent out.
+
```

Comment:
> Should we highlight whether we are making any copies?

It is already implicitly there by saying `takes ownership of the data`.

We could add some more explicit language here, but I'm inclined not to do it.  I don't want to get too bogged down in zero-copy discussions, because we clearly aren't.